service/common/src/main/java/org/apache/polaris/service/persistence/MetadataCacheManager.java

```diff
@@ -0,0 +1,155 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.persistence;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.polaris.core.PolarisCallContext;
+import org.apache.polaris.core.config.FeatureConfiguration;
+import org.apache.polaris.core.entity.PolarisEntity;
+import org.apache.polaris.core.entity.PolarisEntityConstants;
+import org.apache.polaris.core.entity.PolarisEntitySubType;
+import org.apache.polaris.core.entity.PolarisEntityType;
+import org.apache.polaris.core.entity.table.IcebergTableLikeEntity;
+import org.apache.polaris.core.entity.table.TableLikeEntity;
+import org.apache.polaris.core.persistence.PolarisMetaStoreManager;
+import org.apache.polaris.core.persistence.PolarisResolvedPathWrapper;
+import org.apache.polaris.core.persistence.dao.entity.EntityResult;
+import org.apache.polaris.core.persistence.resolver.PolarisResolutionManifestCatalogView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Contains utility methods related to storing TableMetadata in the metastore and retrieving it from
+ * the metastore
+ */
+public class MetadataCacheManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetadataCacheManager.class);
+
+  /**
+   * Load the cached metadata.json content and location or fall back to `fallback` if one doesn't
+   * exist. If the metadata is not currently cached, it may be added to the cache.
+   */
+  public static MetadataJson loadTableMetadataJson(
+      TableIdentifier tableIdentifier,
+      int maxBytesToCache,
+      PolarisCallContext callContext,
+      PolarisMetaStoreManager metastoreManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView,
+      Supplier<TableMetadata> fallback) {
+    LOGGER.debug(String.format("Loading cached metadata for %s", tableIdentifier));
+    PolarisResolvedPathWrapper resolvedEntities =
+        resolvedEntityView.getResolvedPath(
+            tableIdentifier, PolarisEntityType.TABLE_LIKE, PolarisEntitySubType.ICEBERG_TABLE);
+    // If the table doesn't exist, just fall back fast
+    if (resolvedEntities == null) {
+      return MetadataJson.fromMetadata(fallback.get());
+    }
+    IcebergTableLikeEntity tableLikeEntity =
+        IcebergTableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+    String cacheContent = tableLikeEntity.getMetadataCacheContent();
+    if (cacheContent != null) {
+      LOGGER.debug(String.format("Using cached metadata for %s", tableIdentifier));
+      Map<String, String> entityProperties = tableLikeEntity.getPropertiesAsMap();
+      return new MetadataJson(
+          tableLikeEntity.getMetadataLocation(),
+          tableLikeEntity.getMetadataCacheContent(),
+          Stream.of(
```

Comment:
I'm probably out of touch here, but why do these properties need to be added? If there are specific to Polaris's internal representation it seems like we might want to move this into a helper method

getPolarisTableProperties?

Reply:
This is gone now that I've pulled back the StringLoadTableResponse changes